integration/fuse/src/main/java/alluxio/fuse/AlluxioJniFuseFileSystem.java

```diff
@@ -0,0 +1,633 @@
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.fuse;
+
+import alluxio.AlluxioURI;
+import alluxio.client.file.FileInStream;
+import alluxio.client.file.FileOutStream;
+import alluxio.client.file.FileSystem;
+import alluxio.client.file.URIStatus;
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.grpc.CreateDirectoryPOptions;
+import alluxio.grpc.CreateFilePOptions;
+import alluxio.grpc.SetAttributePOptions;
+import alluxio.jnifuse.AbstractFuseFileSystem;
+import alluxio.jnifuse.ErrorCodes;
+import alluxio.jnifuse.FuseFillDir;
+import alluxio.jnifuse.struct.FileStat;
+import alluxio.jnifuse.struct.FuseContext;
+import alluxio.jnifuse.struct.FuseFileInfo;
+import alluxio.resource.LockResource;
+import alluxio.security.authorization.Mode;
+import alluxio.util.ThreadUtils;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Main FUSE implementation class.
+ * <p>
+ * Implements the FUSE callbacks defined by jni-fuse.
+ */
+@ThreadSafe
+public final class AlluxioJniFuseFileSystem extends AbstractFuseFileSystem {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioJniFuseFileSystem.class);
+  private final FileSystem mFileSystem;
+  private final AlluxioConfiguration mConf;
+  // base path within Alluxio namespace that is used for FUSE operations
+  // For example, if alluxio-fuse is mounted in /mnt/alluxio and mAlluxioRootPath
+  // is /users/foo, then an operation on /mnt/alluxio/bar will be translated on
+  // an action on the URI alluxio://<master>:<port>/users/foo/bar
+  private final Path mAlluxioRootPath;
+  // Keeps a cache of the most recently translated paths from String to Alluxio URI
+  private final LoadingCache<String, AlluxioURI> mPathResolverCache;
+  private final LoadingCache<String, Long> mUidCache;
+  private final LoadingCache<String, Long> mGidCache;
+  private final AtomicLong mNextOpenFileId = new AtomicLong(0);
+  private final String mFsName;
+
+  private static final int LOCK_SIZE = 2048;
+  /** A readwrite lock pool to guard individual files based on striping. */
+  private final ReadWriteLock[] mFileLocks = new ReentrantReadWriteLock[LOCK_SIZE];
```

Comment:
This feels like a striped lock? Would this work? https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Striped.html

I know we are using this in other parts of the codebase.

Reply:
good to know! updated to Striped